blueflood-core/src/main/java/com/rackspacecloud/blueflood/outputs/handlers/RollupHandler.java

```diff
@@ -287,24 +304,109 @@ private Boolean repairMetrics (Locator locator, MetricData metricData, final lon
         return retValue;
     }
 
-    private List<Points.Point> repairRollupsOnRead(Locator locator, Granularity g, long from, long to) {
+    /**
+     * This method gets the points from the DB and then rolls them up according to the granularity.
+     *
+     * Breaks up the number of ranges into buckets based on ROLLUP_ON_READ_REPAIR_SIZE_PER_THREAD and executes
+     * the buckets in parallel.
+     *
+     * @param locator metric key within the DB
+     * @param g the granularity
+     * @param from the starting timestamp of the range (ms)
+     * @param to the ending timestamp of the range (ms)
+     *
+     * @return a list of rolled-up points
+     */
+    private List<Points.Point> repairRollupsOnRead(final Locator locator, Granularity g, long from, long to) {
+        Timer.Context c = timerRepairRollupsOnRead.time();
+
+        List<Points.Point> repairedPoints = new ArrayList<Points.Point>();
+        List<ListenableFuture<List<Points.Point>>> futures = new ArrayList<ListenableFuture<List<Points.Point>>>();
+
+        for( final Iterable<Range> ranges : divideRangesByGroup( g, from, to ) ) {
+            futures.add(
+
+                    createRepairPointsExecutor.submit( new Callable() {
+
+                        @Override
+                        public List<Points.Point> call() throws Exception {
+                            return createRepairPoints( ranges, locator );
+                        }
+                    } ) );
+        }
+
+        ListenableFuture<List<List<Points.Point>>> aggregateFuture = Futures.allAsList(futures);
+
+        try {
+            for( List<Points.Point> subList : aggregateFuture.get(rollupOnReadTimeout.getValue(), rollupOnReadTimeout.getUnit()) ) {
+
+                repairedPoints.addAll( subList );
+            }
+        } catch (Exception e) {
+            log.warn("Exception encountered while doing rollups on read, incomplete rollups will be returned.", e);
+        }
+
+        c.stop();
+
+        return repairedPoints;
+    }
+
+    /**
+     * Create a list of groups of ranges based on the to, from, granularity, and ROLLUP_ON_READ_REPAIR_SIZE_PER_THREAD.
+     *
+     * @param g the granularity
+     * @param from the starting timestamp of the range (ms)
+     * @param to the ending timestamp of the range (ms)
+     *
+     * @return A list of iterables of ranges, each iterable list is no bigger than ROLLUP_ON_READ_REPAIR_SIZE_PER_THREAD
+     */
+    private List<Iterable<Range>> divideRangesByGroup( Granularity g, long from, long to ) {
+        List<Iterable<Range>> listRange = new ArrayList<Iterable<Range>>();
+
+        long rangeSize = ROLLUP_ON_READ_REPAIR_SIZE_PER_THREAD * g.milliseconds();
+
+        for ( long start = g.snapMillis( from ); start < to ; ) {
+
+            long end = (start + rangeSize) < to ? (start + rangeSize ) : to;
```

Comment:
If I am understanding this correctly, If you convert both "from" and "to" their corresponding snapMillis, you dont have to do this logic right? You dont have to do it. Just wondering.